Create R script to replicate App outputs (#467)

Gero1999 opened the pull request:
## Issue

Closes #467 

## Description
Implemented the script file. It was made in order to eventually also be used using a settings file

## Definition of Done

- [x] R Script export
- [x] Can be run independently and give same results as in app
- [x] Using aNCA functions

## How to test
Do your choices in the App and export the ZIP folder. In the subfolder find the `/code/session_script.R` and run it. Compare the script outputs with the result files in the exported ZIP

## Contributor checklist
- [x] Code passes lintr checks
- [x] Code passes all unit tests
- [x] New logic covered by unit tests
- [x] New logic is documented
- [x] App or package changes are reflected in NEWS
- [x] Package version is incremented

## Notes to reviewer
- The tests for the function `calculate_ratios_app` will be included here #287 independently 
- The tests for the function `get_session_script_code` has probably to be also done as an e2e test. To indeed make sure that both session and script are matching

m-kolomanski commented on R/get_session_script_code.R:
```diff
@@ -0,0 +1,108 @@
+      "Did you accidentally modify the original template?"
+    )
+    return(invisible(output_path))
+  }
```
suggestion: IMHO this should be a straight error or a warning that returns NULL stating that the function failed to parse the template - if the template was meddled with, I don't think we want to include it.

m-kolomanski commented on R/get_session_script_code.R:
```diff
@@ -0,0 +1,108 @@
+  }
+
+  # Replace all matches with deparsed values in one pass
+  match_positions <- gregexpr(pattern, script, perl = TRUE)[[1]]
```
issue: We already define `matches` this is redundant.

m-kolomanski commented on R/get_session_script_code.R:
```diff
@@ -0,0 +1,108 @@
+    # Multi-line, indented, no trailing comma on last column
+    cols <- lapply(obj, function(col) {
+      d <- paste(deparse(col, width.cutoff = 500), collapse = "")
+      if (grepl("^c\\(", d)) d else paste0("c(", d, ")")
```
style: IMHO it is a bit hard to follow, something like this is less concise but I think a bit more readable:
```suggestion
      d <- col |>
        deparse(width.cutoff = 500) |>
        paste(collapse = "")
        
      if (!grepl("^c\\(", d)) {
        d <- paste0("c(", d, ")")
      }
      
      d
```

m-kolomanski commented on R/get_session_script_code.R:
issue: We definitely need unit tests for `clean_deparse` to make sure we are doing things right.

js3110 reviewed (commented):
Its working! But in the pivoted_results, I have 4 rows for each results row- maybe something is going wrong with some merging at some point?

<img width="1131" height="246" alt="image" src="https://github.com/user-attachments/assets/35abbdcd-4a7f-4022-9097-f1edba2cc7e1" />

js3110 commented on inst/shiny/www/templates/script_template.R:
```diff
@@ -0,0 +1,118 @@
+
+## Setup NCA settings in the PKNCA object ########################
+auc_data <- session$userData$settings$partial_aucs
+units_table <- session$userData$final_units
```
I tried this with the multispec-data but the units table was so long that it wasn't fully expanded and therefore the ) at the end of the line wasn't found and so I couldn't run the code.
Is there a way to set a limit for how long the line of code is before it starts to overflow?

js3110 commented on inst/shiny/www/templates/script_template.R:
```diff
@@ -0,0 +1,125 @@
+library(dplyr)
+
+# Load raw data #
+data_path <- "../data/data.rds"
```
Nitpick: Should add a comment here that makes it clear that this part needs to be modified by the user, or comment it out so its obvious

Gero1999 replied:
this path should already work if they open the file and don't change the directory. Otherwise they will need to change the directory as indicated in the previous comment

js3110 replied:
ah, i see, youre right. thanks!

js3110 reviewed (approved):
LGTM, some non-blocking comments.

js3110 reviewed (commented):
I am still getting this message every time I checkout your branch:

```
Warning message:
── Conflicts ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── aNCA conflicts
──
✖ `calculate_table_ratios_app` masks `aNCA::calculate_table_ratios_app()`.
ℹ Did you accidentally source a file rather than using `load_all()`?
  Run `rm(list = c("calculate_table_ratios_app"))` to remove the conflicts. 
```
  
  the only way to solve is to clear my env before I run load_all()


In addition, I have a similar issue to @Shaakon35 : the PK Params in the r script version are more than in the output results. I investigated a little and I think its because in the app we filter out the params calculated due to dependencies but not requested, but in the rscript its not filtering them.

Gero1999 commented:
> At the very least, if we want to implement it this way right now and then add the functionality of parsing a settings file instead of session object, we should have: more generic name, TODO comments

I created a TODO and the implementation will be developed in #835. The name/refactoring of this function I will be done there just for easiness.

Here, in this PR, I removed the redundant check, renamed `calculate_table_ratios_app` > `calculate_table_ratios` & refactored to use `apply`

Shaakon35 reviewed (approved):
Here we go!! Congratulations @Gero1999 👍

Shaakon35 commented:
@Gero1999 Can you add in the README, in the Description bullet list:
- reproduce your NCA outside of the app, with the R script created

m-kolomanski commented on R/pivot_wider_pknca_results.R:
```diff
@@ -170,3 +204,60 @@ pivot_wider_pknca_results <- function(myres) {
+  # Add flaging columns in the pivoted results
+  applied_flags <- purrr::keep(flag_rules, function(x) x$is.checked)
+  flag_params <- names(flag_rules)
+  flag_thr <- sapply(flag_rules, FUN =  function(x) x$threshold)
```
nitpick:
```suggestion
  flag_thr <- sapply(flag_rules, FUN =  \(x) x$threshold)
```

also consider `purrr::map_chr()`
